File: data_structures/segment_tree/matlab/segment_tree.m
function varargout = segment_tree(varargin)
%SEGMENT_TREE  セグメント木（区間和）
%
%   st = SEGMENT_TREE('new', data)          配列からセグメント木を構築
%   st = SEGMENT_TREE(st, 'update', i, v)  i番目（1始まり）を更新
%   s  = SEGMENT_TREE(st, 'query', l, r)   [l, r]（1始まり）の区間和
%   n  = SEGMENT_TREE(st, 'size')          要素数

    if nargin >= 2 && ischar(varargin{1}) && strcmp(varargin{1}, 'new')
        data = varargin{2};
        n = length(data);
        tree = zeros(1, 4 * n);
        st = struct('tree', tree, 'n', n);
        if n > 0
            st = seg_build(st, data, 1, 1, n);
        end
        varargout{1} = st;
        return;
    end

    st = varargin{1};
    op = varargin{2};

    switch op
        case 'update'
            i = varargin{3};
            v = varargin{4};
            if i < 1 || i > st.n
                error('segment_tree:outOfRange', 'index out of range');
            end
            st = seg_update(st, 1, 1, st.n, i, v);
            varargout{1} = st;

        case 'query'
            l = varargin{3};
            r = varargin{4};
            if l < 1 || r > st.n || l > r
                error('segment_tree:outOfRange', 'invalid range');
            end
            varargout{1} = seg_query(st, 1, 1, st.n, l, r);

        case 'size'
            varargout{1} = st.n;

        otherwise
            error('segment_tree:unknownOp', 'Unknown operation: %s', op);
    end
end

function st = seg_build(st, data, node, lo, hi)
    if lo == hi
        st.tree(node) = data(lo);
        return;
    end
    mid = floor((lo + hi) / 2);
    st = seg_build(st, data, 2*node, lo, mid);
    st = seg_build(st, data, 2*node+1, mid+1, hi);
    st.tree(node) = st.tree(2*node) + st.tree(2*node+1);
end

function st = seg_update(st, node, lo, hi, i, v)
    if lo == hi
        st.tree(node) = v;
        return;
    end
    mid = floor((lo + hi) / 2);
    if i <= mid
        st = seg_update(st, 2*node, lo, mid, i, v);
    else
        st = seg_update(st, 2*node+1, mid+1, hi, i, v);
    end
    st.tree(node) = st.tree(2*node) + st.tree(2*node+1);
end

function s = seg_query(st, node, lo, hi, l, r)
    if r < lo || hi < l, s = 0; return; end
    if l <= lo && hi <= r, s = st.tree(node); return; end
    mid = floor((lo + hi) / 2);
    s = seg_query(st, 2*node, lo, mid, l, r) + seg_query(st, 2*node+1, mid+1, hi, l, r);
end

File: data_structures/segment_tree/cpp/segment_tree.cpp
#include "segment_tree.h"

SegmentTree::SegmentTree(const std::vector<int>& data)
    : n_(static_cast<int>(data.size())), tree_(4 * n_, 0) {
    if (n_ > 0) build(data, 1, 0, n_ - 1);
}

void SegmentTree::build(const std::vector<int>& data, int node, int lo, int hi) {
    if (lo == hi) { tree_[node] = data[lo]; return; }
    int mid = (lo + hi) / 2;
    build(data, 2 * node, lo, mid);
    build(data, 2 * node + 1, mid + 1, hi);
    tree_[node] = tree_[2 * node] + tree_[2 * node + 1];
}

void SegmentTree::update(int node, int lo, int hi, int i, int value) {
    if (lo == hi) { tree_[node] = value; return; }
    int mid = (lo + hi) / 2;
    if (i <= mid) update(2 * node, lo, mid, i, value);
    else          update(2 * node + 1, mid + 1, hi, i, value);
    tree_[node] = tree_[2 * node] + tree_[2 * node + 1];
}

void SegmentTree::update(int i, int value) {
    if (i < 0 || i >= n_) throw std::out_of_range("index out of range");
    update(1, 0, n_ - 1, i, value);
}

long long SegmentTree::query(int node, int lo, int hi, int l, int r) const {
    if (r < lo || hi < l) return 0;
    if (l <= lo && hi <= r) return tree_[node];
    int mid = (lo + hi) / 2;
    return query(2 * node, lo, mid, l, r) + query(2 * node + 1, mid + 1, hi, l, r);
}

long long SegmentTree::query(int l, int r) const {
    if (l < 0 || r > n_ || l >= r) throw std::out_of_range("invalid range");
    return query(1, 0, n_ - 1, l, r - 1);
}

int SegmentTree::size() const { return n_; }
